crates/sui-display/src/v2/lexer.rs

```diff
@@ -0,0 +1,416 @@
+/// each containing nested field, vector, or dynamic field accesses.
+#[derive(Debug)]
+pub(crate) struct Lexer<'s> {
+    /// Remaining input to be tokenized.
+    src: &'s str,
+
+    /// The number of bytes (not characters) tokenized so far.
+    off: usize,
+
+    /// Whether the lexer is currently inside a text strand or an expression strand.
+    mode: Mode,
+}
+
+#[derive(Debug)]
+enum Mode {
+    Text,
+    Expr,
+}
+
+/// A lexeme is a token along with its offset in the source string, and the slice of source string
+/// that it originated from.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) struct Lexeme<'s>(pub Token, pub usize, pub &'s str);
+
+/// Like [Lexeme] but owns the slice of source string. Useful for capturing context in an error
+/// message.
+#[derive(Debug)]
+pub(crate) struct OwnedLexeme(Token, usize, String);
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) enum Token {
+    /// '.'
+    Dot,
+    /// An identifier
+    Ident,
+    /// '{'
+    LBrace,
+    /// '{{'
+    LLBrace,
+    /// '}'
+    RBrace,
+    /// '}}'
+    RRBrace,
+    /// A strand of text.
+    Text,
+
+    /// An unexpected byte in the input string.
+    Unexpected,
+}
+
+#[derive(Debug)]
+pub(crate) struct TokenSet<'t>(pub &'t [Token]);
+
+impl<'s> Lexer<'s> {
+    pub(crate) fn new(src: &'s str) -> Self {
+        Self {
+            src,
+            off: 0,
+            mode: Mode::Text,
+        }
+    }
+
+    /// Assuming the lexer is in text mode, return the next text token.
+    fn next_text_token(&mut self) -> Option<Lexeme<'s>> {
+        let bytes = self.src.as_bytes();
+
+        use Token as T;
+        Some(match bytes.first()? {
+            b'{' if bytes.get(1) == Some(&b'{') => {
+                self.take(T::LLBrace, 1); // discard the first brace
+                self.take(T::LLBrace, 1)
+            }
+
+            b'{' => {
+                self.mode = Mode::Expr;
+                self.take(T::LBrace, 1)
+            }
+
+            b'}' if bytes.get(1) == Some(&b'}') => {
+                self.take(T::RRBrace, 1); // discard the first brace
+                self.take(T::RRBrace, 1)
+            }
+
+            // This is not a valid token within text, but recognise it so that the parser can
+            // produce a better error message.
+            b'}' => self.take(T::RBrace, 1),
+
+            _ => self.take_until(T::Text, |c| ['{', '}'].contains(&c)),
+        })
+    }
+
+    /// Assuming the lexer is in expression mode, return the next expression token.
+    fn next_expr_token(&mut self) -> Option<Lexeme<'s>> {
+        self.skip_whitespace();
+
+        use Token as T;
+        Some(match self.src.as_bytes().first()? {
+            b'{' => self.take(T::LBrace, 1),
+
+            b'}' => {
+                self.mode = Mode::Text;
+                self.take(T::RBrace, 1)
+            }
+
+            b'.' => self.take(T::Dot, 1),
+
+            b'a'..=b'z' | b'A'..=b'Z' => {
```

Comment:
Maybe I'm missing the place (apologies if so!) or I'm forgetting some discussion, but this doesn't encompass all valid identifiers still AFAICT (e.g., `____`)

Reply:
ah! It's in the PR above this one :)